src/Form/FormProtector.php

```diff
@@ -0,0 +1,605 @@
+            if (array_values(array_intersect($field, $unlockParts)) === $unlockParts) {
+                return $this;
+            }
+        }
+
+        $field = implode('.', $field);
+        $field = preg_replace('/(\.\d+)+$/', '', $field);
+
+        if ($lock) {
+            if (!in_array($field, $this->fields, true)) {
+                if ($value !== null) {
+                    $this->fields[$field] = $value;
+
+                    return $this;
+                }
+                if (isset($this->fields[$field])) {
+                    unset($this->fields[$field]);
+                }
+                $this->fields[] = $field;
+            }
+        } else {
+            $this->unlockField($field);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Parses the field name to create a dot separated name value for use in
+     * field hash. If filename is of form Model[field] or Model.fiedl an array of
+     * fieldname parts like ['Model', 'field'] is returned.
+     *
+     * @param string $name The form inputs name attribute.
+     * @return string[] Array of field name params like ['Model.field'] or
+     *   ['Model', 'field'] for array fields or empty array if $name is empty.
+     */
+    protected function getFieldNameArray(string $name): array
+    {
+        if (empty($name) && $name !== '0') {
+            return [];
+        }
+
+        if (strpos($name, '[') === false) {
+            return Hash::filter(explode('.', $name));
+        }
+        $parts = explode('[', $name);
+        $parts = array_map(function ($el) {
+            return trim($el, ']');
+        }, $parts);
+
+        return Hash::filter($parts, 'strlen');
+    }
+
+    /**
+     * Add to the list of fields that are currently unlocked.
+     *
+     * Unlocked fields are not included in the field hash.
+     *
+     * @param string $name The dot separated name for the field.
+     * @return $this
+     */
+    public function unlockField($name)
+    {
+        if (!in_array($name, $this->unlockedFields, true)) {
+            $this->unlockedFields[] = $name;
+        }
+
+        $index = array_search($name, $this->fields, true);
+        if ($index !== false) {
+            unset($this->fields[$index]);
+        }
+        unset($this->fields[$name]);
+
+        return $this;
+    }
+
+    /**
+     * Get validation error message.
+     *
+     * @return string|null
+     */
+    public function getError(): ?string
+    {
+        return $this->debugMessage;
+    }
+
+    /**
+     * Extract token from data.
+     *
+     * @param mixed $formData Data to validate.
+     * @return string|null Fields token on success, null on failure.
+     */
+    protected function extractToken($formData): ?string
+    {
+        if (!is_array($formData)) {
+            $this->debugMessage = 'Request data is not an array.';
+
+            return null;
+        }
+
+        $message = '\'%s\' was not found in request data.';
```

Comment:
```suggestion
        $message = '`%s`' was not found in request data.';
```
Will get formatted better by the development error pages.